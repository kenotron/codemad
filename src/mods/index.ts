export { renameJsxProp } from './renameJsxProp';
export { replaceJsxPropValue } from './replaceJsxPropValue';
