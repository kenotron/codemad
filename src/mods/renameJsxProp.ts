import ts from 'typescript';
import { Modder } from '../interfaces';

export function renameJsxProp(tagName: string, propFrom: string, propTo: string) {
  return function renameJsxPropVisitor(node: ts.Node, modder: Modder) {
    if (ts.isJsxOpeningLikeElement(node)) {
      if (node.tagName.getText() === tagName && node.attributes && node.attributes.properties) {
        for (let prop of node.attributes.properties) {
          if (ts.isJsxAttribute(prop) && prop && prop.name && prop.name.getText() === propFrom && prop.initializer) {
            return modder.replace(prop.name, propTo);
          }
        }
      }
    }
  };
}
