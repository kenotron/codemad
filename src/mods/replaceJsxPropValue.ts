import ts from 'typescript';
import { Modder } from '../interfaces';

export function replaceJsxPropValue(tagName: string, propName: string, value: string) {
  return function replaceJsxPropValueVisitor(node: ts.Node, modder: Modder) {
    if (ts.isJsxOpeningLikeElement(node)) {
      if (node.tagName.getText() === tagName && node.attributes && node.attributes.properties) {
        for (let prop of node.attributes.properties) {
          if (ts.isJsxAttribute(prop) && prop && prop.name && prop.name.getText() === propName && prop.initializer) {
            return modder.replace(prop.initializer, value);
          }
        }
      }
    }

    return undefined;
  };
}
