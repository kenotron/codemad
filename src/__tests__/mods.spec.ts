import mockfs from 'mock-fs';
import { renameJsxProp } from '../mods/renameJsxProp';
import { replaceJsxPropValue } from '../mods/replaceJsxPropValue';
import { loadFixtures } from './testutils/loadFixtures';
import { testTsHandlerWithFixtureFactory } from './testutils/testTsHandlerWithFixtureFactory';

const FixturesPath = 'src/__tests__/fixtures/mods';

describe('typescript mods', () => {
  const mockContents = loadFixtures(FixturesPath);
  const testWithFixture = testTsHandlerWithFixtureFactory(FixturesPath);

  beforeEach(() => {
    mockfs(mockContents);
  });

  it('replaces a single prop name in jsx', () => {
    testWithFixture('renameJsxProp.tsx', {}, renameJsxProp('App', 'hello', 'moddedProp'));
  });

  it('replaces a single prop value in jsx', () => {
    testWithFixture('replaceJsxPropValue.tsx', {}, replaceJsxPropValue('App', 'hello', '"modded"'));
  });
});