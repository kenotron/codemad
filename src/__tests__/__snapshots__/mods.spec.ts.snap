// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typescript mods replaces a single prop name in jsx 1`] = `
"import React from 'react';
import ReactDOM from 'react-dom';

const App = (props: any) => {
  return <div>{props.hello}</div>;
};

ReactDOM.render(<App moddedProp=\\"world\\" />, document.createElement('div'));
"
`;

exports[`typescript mods replaces a single prop value in jsx 1`] = `
"import React from 'react';
import ReactDOM from 'react-dom';

const App = (props: any) => {
  return <div>{props.hello}</div>;
};

ReactDOM.render(<App hello=\\"modded\\" />, document.createElement('div'));
"
`;
